test/e2e/envvars_test.go

```diff
@@ -81,18 +83,78 @@ func TestEnvVars(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Error fetching Route %s: %v", names.Route, err)
 	}
-	domain := route.Status.Domain
 
-	envVarsExpectedOutput := fmt.Sprintf("Here are our env vars service:%s - configuration:%s - revision:%s", names.Service, names.Config, names.Revision)
+	domain := route.Status.Domain + "/envvars/should"
 
-	_, err = pkgTest.WaitForEndpointState(
+	resp, err := pkgTest.WaitForEndpointState(
 		clients.KubeClient,
 		logger,
 		domain,
-		pkgTest.Retrying(pkgTest.MatchesBody(envVarsExpectedOutput), http.StatusNotFound),
+		pkgTest.Retrying(verifyUpStatus, http.StatusNotFound),
 		"EnvVarsServesText",
 		test.ServingFlags.ResolvableDomain)
 	if err != nil {
-		t.Fatalf("The endpoint for Route %s at domain %s didn't serve the expected text \"%s\": %v", names.Route, domain, envVarsExpectedOutput, err)
+		t.Fatalf("Failed before reaching desired state : %v", err)
+	}
+
+	expectedOutputs := map[string]string{
+		"K_SERVICE" : names.Service,
+		"K_CONFIGURATION" : names.Config,
+		"K_REVISION" : names.Revision,
+	}
+	// We are failing even for "SHOULD".
+	if err = verifyResponse(resp, "SHOULD", expectedOutputs); err != nil {
+		t.Error(err)
+	}
+
+	domain = route.Status.Domain + "/envvars/must"
+
+	resp, err = pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		logger,
+		domain,
+		pkgTest.Retrying(verifyUpStatus, http.StatusNotFound),
+		"EnvVarsServesText",
+		test.ServingFlags.ResolvableDomain)
+	if err != nil {
+		t.Fatalf("Failed before reaching desired state : %v", err)
+	}
+
+	expectedOutputs = map[string]string{
+		"PORT" : "8080",
+	}
+	if err = 	verifyResponse(resp, "MUST", expectedOutputs); err != nil {
+		t.Error(err)
 	}
 }
+
+func verifyUpStatus(resp *spoof.Response) (done bool, err error) {
+	if resp.StatusCode ==  http.StatusOK {
+		return true, nil
+	}
+
+	return true, errors.New(string(resp.Body))
+}
+
+func verifyResponse(resp *spoof.Response, constraint string, expectedOutputs map[string]string) error {
+	var respMap map[string]string
+	if jsonErr := json.Unmarshal(resp.Body, &respMap); jsonErr != nil {
+		return errors.Errorf("Json Un-marshalling error : %v", jsonErr)
+	}
+
+	if len(expectedOutputs) != len(respMap) {
```

Comment:
I'd tend to lean towards a more permissive parsing model here. In particular, this test ends up spreading the should/must logic across both the test itself and the running container. It might be simpler to do the following:

1. Have the target container simply return a JSON-encoded envvars map (object).
2. Have the test decode the JSON with a "should" and a "must" struct, like so:

```golang

var should struct {
  service string `json:"K_SERVICE"`
  configuration string `json:"K_CONFIGURATION"`
  revision string `json:"K_REVISION"`
}
var must struct {
  port string `json:"PORT"`
}

if err := json.Unmarshal(payload, &should); err != nil {
  ...
}
if err := json.Unmarshal(payload, &must); err != nil {
  ...
}

if should != struct{...} {
  ...
}
if must != struct{...} {
 ...
}
```

Reply:
Makes sense, I have updated the verification to use this model.